server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletTransactionLog.java

```diff
@@ -0,0 +1,549 @@
+    public static final int OVERRUN_THRESHOLD = (Integer.MAX_VALUE / 2);
+    public static final int MAX_ALLOWED_SIZE = (OVERRUN_THRESHOLD / 2 - 1);
+
+    /**
+     * Create a new ring with the specified capacity
+     *
+     * @param capacity The capacity of the ring
+     * @throws IllegalArgumentException if the size is too large
+     */
+    public Ring(int capacity) {
+      if (capacity > MAX_ALLOWED_SIZE) {
+        throw new IllegalArgumentException("Size cannot be larger than " + MAX_ALLOWED_SIZE);
+      }
+      ring = new Object[capacity];
+      first = 0;
+      last = -1;
+    }
+
+    /**
+     * Add an item to the ring
+     *
+     * @param object The object to add
+     * @return the item removed from the back of the ring if any
+     */
+    @SuppressWarnings("unchecked")
+    public T add(T object) {
+      Object removed = null;
+      if (ring.length > 0) {
+        if (size() == ring.length) {
+          removed = ring[first++ % ring.length];
+        }
+        int newEnd = last + 1;
+        ring[newEnd % ring.length] = object;
+        last = newEnd;
+      } else {
+        removed = object;
+      }
+      avoidOverrun();
+      return (T) removed;
+    }
+
+    /**
+     * The size of the ring
+     *
+     * @return the size
+     */
+    public int size() {
+      return last - first + 1;
+    }
+
+    /**
+     * The capacity of the ring
+     *
+     * @return the capacity
+     */
+    public int capacity() {
+      return ring.length;
+    }
+
+    /**
+     * Is the ring empty
+     *
+     * @return true if the ring is empty
+     */
+    public boolean isEmpty() {
+      return last < first;
+    }
+
+    /**
+     * Clear/empty the ring.
+     */
+    public void clear() {
+      last = first - 1;
+    }
+
+    /**
+     * This is the only method that can be done concurrently with a modification via the add method.
+     * This will continue to call toList as long as a concurrent modification exception is thrown.
+     *
+     * @return The list of objects in the ring.
+     */
+    public List<T> toListNoFail() {
+      while (true) {
+        try {
+          return toList();
+        } catch (ConcurrentModificationException cme) {
+          // try again.
+        }
+      }
+    }
+
+    /**
+     * Return the list of objects in the ring.
+     *
+     * @return the list of objects
+     * @throws ConcurrentModificationException if the list is modified while gathering the list
+     */
+    @SuppressWarnings("unchecked")
+    public List<T> toList() throws ConcurrentModificationException {
+      Object[] data = null;
+      long updateCount = getUpdateCount();
+      int lastPos = last;
+      int firstPos = first;
+      // if either is over the threshold, then we must be in the middle of a modification
+      // but before the overrun threshold is called. try again.
+      if (lastPos <= OVERRUN_THRESHOLD && firstPos <= OVERRUN_THRESHOLD) {
+        data = new Object[lastPos - firstPos + 1];
```

Comment:
If last is read and than other threads increment first a bunch before its read then could possibly try to create an array with a negative size.  Added the following to prevent this.

Do not see how the `size <= ring.length` case could ever not be true, but also did not prove it could not happen to myself so added it.

```suggestion
      int size = lastPos - firstPos + 1;
      if (size >= 0 && size <= ring.length && lastPos <= OVERRUN_THRESHOLD
          && firstPos <= OVERRUN_THRESHOLD) {
        data = new Object[size];

```

Reply:
well, safety first I guess.  However I think I proved to myself that a negative size here cannot happen.